pkg/koordlet/metricsadvisor/collectors/coldmemoryresource/cold_page_kidled.go

```diff
@@ -0,0 +1,237 @@
+		}
+		system.SetIsStartColdMemory(true)
+		return true
+	}
+	return false
+}
+
+func (k *kidledcoldPageCollector) Setup(c1 *framework.Context) {}
+
+func (k *kidledcoldPageCollector) collectColdPageInfo() {
+	if k.statesInformer == nil {
+		return
+	}
+	coldPageMetrics := make([]metriccache.MetricSample, 0)
+
+	nodeColdPageInfoMetric, err := k.collectNodeColdPageInfo()
+	if err != nil {
+		klog.Warningf("generate node cold page info metrics failed, err %v", err)
+	}
+	coldPageMetrics = append(coldPageMetrics, nodeColdPageInfoMetric...)
+
+	podsColdPageInfoMetric, err := k.collectPodsColdPageInfo()
+	if err != nil {
+		klog.Warningf("generate pods or container cold page info metrics failed, err %v", err)
+	}
+	coldPageMetrics = append(coldPageMetrics, podsColdPageInfoMetric...)
+
+	appender := k.appendableDB.Appender()
+	if err := appender.Append(coldPageMetrics); err != nil {
+		klog.ErrorS(err, "Append node metrics error")
+		return
+	}
+
+	if err := appender.Commit(); err != nil {
+		klog.Warningf("Commit node metrics failed, reason: %v", err)
+		return
+	}
+
+	k.started.Store(true)
+}
+
+func (k *kidledcoldPageCollector) collectNodeColdPageInfo() ([]metriccache.MetricSample, error) {
+	coldPageMetrics := make([]metriccache.MetricSample, 0)
+	collectTime := time.Now()
+	nodeColdPageBytes, err := k.cgroupReader.ReadMemoryColdPageUsage("")
+	if err != nil {
+		return nil, err
+	}
+	nodeColdPageBytesValue := float64(nodeColdPageBytes)
+	nodeColdPageMetrics, err := metriccache.NodeMemoryColdPageSizeMetric.GenerateSample(nil, collectTime, nodeColdPageBytesValue)
+	if err != nil {
+		return nil, err
+	}
+	coldPageMetrics = append(coldPageMetrics, nodeColdPageMetrics)
+
+	memUsageWithHotPageBytes, err := koordletutil.GetNodeMemUsageWithHotPage(nodeColdPageBytes)
+	if err != nil {
+		return nil, err
+	}
+	memUsageWithHotPageValue := float64(memUsageWithHotPageBytes)
+	memUsageWithHotPageMetrics, err := metriccache.NodeMemoryWithHotPageUsageMetric.GenerateSample(nil, collectTime, memUsageWithHotPageValue)
+	if err != nil {
+		return nil, err
+	}
+	coldPageMetrics = append(coldPageMetrics, memUsageWithHotPageMetrics)
+	klog.V(4).Infof("collectNodeResUsed finished, count %v, memUsageWithHotPage[%v], coldPageSize[%v]",
+		len(coldPageMetrics), memUsageWithHotPageValue, nodeColdPageBytes)
+	return coldPageMetrics, nil
+}
+
+func (k *kidledcoldPageCollector) collectPodsColdPageInfo() ([]metriccache.MetricSample, error) {
+	podMetas := k.statesInformer.GetAllPods()
+	count := 0
+	coldMetrics := make([]metriccache.MetricSample, 0)
+	for _, meta := range podMetas {
+		pod := meta.Pod
+		uid := string(pod.UID) // types.UID
+		podKey := util.GetPodKey(pod)
+		if filtered, msg := k.FilterPod(meta); filtered {
+			klog.V(5).Infof("skip collect pod %s, reason: %s", podKey, msg)
+			continue
+		}
+		collectTime := time.Now()
+		podCgroupDir := meta.CgroupDir
+		podColdPageBytes, err := k.cgroupReader.ReadMemoryColdPageUsage(podCgroupDir)
+		if err != nil {
```

Comment:
should ignore the error for the pod/container whose phase is not running or pending since it an expected failure while collecting a non-running pod/container.
please follow the implementation of the pod resource collector